sql/api/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala

```diff
@@ -80,10 +80,12 @@ object RowEncoder extends DataTypeErrorsBase {
       case DoubleType => BoxedDoubleEncoder
       case dt: DecimalType => JavaDecimalEncoder(dt, lenientSerialization = true)
       case BinaryType => BinaryEncoder
-      case CharType(_) | VarcharType(_) =>
-        throw new AnalysisException(
-          errorClass = "UNSUPPORTED_DATA_TYPE_FOR_ENCODER",
-          messageParameters = Map("dataType" -> toSQLType(dataType)))
+      case CharType(_) | VarcharType(_) if !SqlApiConf.get.preserveCharVarcharTypeInfo =>
+          throw new AnalysisException(
+            errorClass = "UNSUPPORTED_DATA_TYPE_FOR_ENCODER",
+            messageParameters = Map("dataType" -> toSQLType(dataType)))
+      case CharType(length) => CharEncoder(length)
+      case VarcharType(length) => VarcharEncoder(length)
```

Comment:
```suggestion
      case CharType(length) if SqlApiConf.get.preserveCharVarcharTypeInfo => CharEncoder(length)
      case VarcharType(length) if SqlApiConf.get.preserveCharVarcharTypeInfo => VarcharEncoder(length)
      case CharType(_) | VarcharType(_) if !SqlApiConf.get.preserveCharVarcharTypeInfo =>
          throw new AnalysisException(
            errorClass = "UNSUPPORTED_DATA_TYPE_FOR_ENCODER",
            messageParameters = Map("dataType" -> toSQLType(dataType)))
```
Why I suggest this is that this is more error proof and more readable. What you say like this is that default behaviour is throw an error, which is what the behaviour was before. With your implementation I feel like this config was added to make something fail, but the default behaviour was to pass.